src/lexer.py

```diff
@@ -0,0 +1,1060 @@
+		self.context = context
+
+	def as_string(self):
+		result  = self.generate_traceback()
+		result += f'{self.error_name}: {self.details}'
+		result += '\n\n' + string_with_arrows(self.pos_start.ftxt, self.pos_start, self.pos_end)
+		return result
+
+	def generate_traceback(self):
+		result = ''
+		pos = self.pos_start
+		ctx = self.context
+
+		while ctx:
+			result = f'  File {pos.fn}, line {str(pos.ln + 1)}, in {ctx.display_name}\n' + result
+			pos = ctx.parent_entry_pos
+			ctx = ctx.parent
+
+		return 'Traceback (most recent call last):\n' + result
+
+#######################################
+# POSITION
+#######################################
+
+class Position:
+	def __init__(self, idx, ln, col, fn, ftxt):
+		self.idx = idx
+		self.ln = ln
+		self.col = col
+		self.fn = fn
+		self.ftxt = ftxt
+
+	def advance(self, current_char=None):
+		self.idx += 1
+		self.col += 1
+
+		if current_char == '\n':
+			self.ln += 1
+			self.col = 0
+
+		return self
+
+	def copy(self):
+		return Position(self.idx, self.ln, self.col, self.fn, self.ftxt)
+
+#######################################
+# TOKENS
+#######################################
+
+TT_INT			= 'INT'
+TT_FLOAT    	= 'FLOAT'
+TT_IDENTIFIER	= 'IDENTIFIER'
+TT_KEYWORD		= 'KEYWORD'
+TT_PLUS     	= 'PLUS'
+TT_MINUS    	= 'MINUS'
+TT_MUL      	= 'MUL'
+TT_DIV      	= 'DIV'
+TT_POW			= 'POW'
+TT_EQ			= 'EQ'
+TT_LPAREN   	= 'LPAREN'
+TT_RPAREN   	= 'RPAREN'
+TT_EE			= 'EE'
+TT_NE			= 'NE'
+TT_LT			= 'LT'
+TT_GT			= 'GT'
+TT_LTE			= 'LTE'
+TT_GTE			= 'GTE'
+TT_EOF			= 'EOF'
+
+KEYWORDS = [
+	'var','and','or','not',
+	'if','elif','else','for',
+	'to','step','while','then'
+]
+
+class Token:
+	def __init__(self, type_, value=None, pos_start=None, pos_end=None):
+		self.type = type_
+		self.value = value
+
+		if pos_start:
+			self.pos_start = pos_start.copy()
+			self.pos_end = pos_start.copy()
+			self.pos_end.advance()
+
+		if pos_end:
+			self.pos_end = pos_end.copy()
+
+	def matches(self, type_, value):
+		return self.type == type_ and self.value == value
+
+	def __repr__(self):
+		if self.value: return f'{self.type}:{self.value}'
+		return f'{self.type}'
+
+#######################################
+# LEXER
+#######################################
+
+class Lexer:
+	def __init__(self, fn, text):
+		self.fn = fn
+		self.text = text
+		self.pos = Position(-1, 0, -1, fn, text)
+		self.current_char = None
+		self.advance()
+
+	def advance(self):
+		self.pos.advance(self.current_char)
+		self.current_char = self.text[self.pos.idx] if self.pos.idx < len(self.text) else None
+
+	def make_tokens(self):
+		tokens = []
+
+		while self.current_char != None:
+			if self.current_char in ' \t':
+				self.advance()
+			elif self.current_char in DIGITS:
+				tokens.append(self.make_number())
+			elif self.current_char in LETTERS:
+				tokens.append(self.make_identifier())
+			elif self.current_char == '+':
+				tokens.append(Token(TT_PLUS, pos_start=self.pos))

```

Comment:
I'm not sure what Token does, but that first parameter, `TT_PLUS` should also be preceded by it's method keyword, like you do with `pos_start=self.pos`

Reply:
does it have a huge effect?